File: python/statecli/file_tracker.py
"""
File Tracker - Auto-tracking of file edits with diffs

Automatically tracks file changes and integrates with StateCLI.
"""

import os
import hashlib
from datetime import datetime
from typing import Optional, List, Dict, Any
from dataclasses import dataclass

from .client import StateCLI


@dataclass
class FileChange:
    """Represents a tracked file change."""
    file_path: str
    operation: str  # 'create', 'modify', 'delete'
    before: Optional[str]
    after: Optional[str]
    diff: List[str]
    timestamp: str


class FileTracker:
    """
    Tracks file edits with automatic diff generation.
    
    Example:
        cli = StateCLI()
        tracker = FileTracker(cli)
        
        # Track a file edit
        change = tracker.track_edit(
            "src/index.py",
            before_content="x = 1",
            after_content="x = 2"
        )
        print(change.diff)  # ['-1: x = 1', '+1: x = 2']
    """
    
    def __init__(self, statecli: StateCLI, auto_checkpoint_threshold: int = 10):
        self.statecli = statecli
        self.auto_checkpoint_threshold = auto_checkpoint_threshold
        self._change_count = 0
    
    def track_edit(
        self,
        file_path: str,
        before_content: str,
        after_content: str,
        actor: str = "ai-agent"
    ) -> FileChange:
        """Track a file edit with before/after content."""
        return self._track_change(file_path, "modify", before_content, after_content, actor)
    
    def track_create(
        self,
        file_path: str,
        content: str,
        actor: str = "ai-agent"
    ) -> FileChange:
        """Track a file creation."""
        return self._track_change(file_path, "create", None, content, actor)
    
    def track_delete(
        self,
        file_path: str,
        previous_content: str,
        actor: str = "ai-agent"
    ) -> FileChange:
        """Track a file deletion."""
        return self._track_change(file_path, "delete", previous_content, None, actor)
    
    def get_file_history(self, file_path: str):
        """Get change history for a file."""
        normalized_path = os.path.normpath(file_path)
        return self.statecli.replay(f"file:{normalized_path}")
    
    def _track_change(
        self,
        file_path: str,
        operation: str,
        before: Optional[str],
        after: Optional[str],
        actor: str
    ) -> FileChange:
        """Internal method to track a file change."""
        diff = self._compute_diff(before, after)
        timestamp = datetime.now().isoformat()
        normalized_path = os.path.normpath(file_path)
        
        change = FileChange(
            file_path=normalized_path,
            operation=operation,
            before=before,
            after=after,
            diff=diff,
            timestamp=timestamp
        )
        
        # Track in StateCLI
        self.statecli.track("file", normalized_path, {
            "operation": operation,
            "content_hash": self._hash_content(after) if after else None,
            "line_count": len(after.split('\n')) if after else 0,
            "diff": diff[:50],  # Store first 50 diff lines
            "timestamp": timestamp
        }, actor)
        
        self._change_count += 1
        
        # Auto-checkpoint if threshold reached
        if self._change_count >= self.auto_checkpoint_threshold:
            self._create_auto_checkpoint()
        
        return change
    
    def _compute_diff(self, before: Optional[str], after: Optional[str]) -> List[str]:
        """Compute simple line-by-line diff."""
        diff = []
        
        if before is None and after is not None:
            # New file
            for i, line in enumerate(after.split('\n'), 1):
                diff.append(f"+{i}: {line}")
        elif before is not None and after is None:
            # Deleted file
            for i, line in enumerate(before.split('\n'), 1):
                diff.append(f"-{i}: {line}")
        elif before is not None and after is not None:
            # Modified file
            before_lines = before.split('\n')
            after_lines = after.split('\n')
            max_lines = max(len(before_lines), len(after_lines))
            
            for i in range(max_lines):
                before_line = before_lines[i] if i < len(before_lines) else None
                after_line = after_lines[i] if i < len(after_lines) else None
                
                if before_line != after_line:
                    if before_line is not None:
                        diff.append(f"-{i+1}: {before_line}")
                    if after_line is not None:
                        diff.append(f"+{i+1}: {after_line}")
        
        return diff
    
    def _hash_content(self, content: str) -> str:
        """Create MD5 hash of content."""
        return hashlib.md5(content.encode()).hexdigest()
    
    def _create_auto_checkpoint(self):
        """Create automatic checkpoint."""
        name = f"auto-{datetime.now().strftime('%Y%m%d-%H%M%S')}"
        self.statecli.checkpoint("session:current", name)
        self._change_count = 0

File: src/file-tracker.ts
/**
 * File Tracker - Auto-tracking of file edits
 * 
 * Automatically tracks file changes and integrates with StateCLI
 */

import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import { StateCLI } from './statecli';

export interface FileChange {
  filePath: string;
  operation: 'create' | 'modify' | 'delete';
  before: string | null;
  after: string | null;
  diff: string[];
  timestamp: string;
}

export interface FileTrackerConfig {
  watchPaths: string[];
  ignorePatterns: string[];
  autoCheckpoint: boolean;
  checkpointThreshold: number; // Number of changes before auto-checkpoint
}

const DEFAULT_CONFIG: FileTrackerConfig = {
  watchPaths: ['.'],
  ignorePatterns: ['node_modules', '.git', 'dist', '*.log', '.statecli'],
  autoCheckpoint: true,
  checkpointThreshold: 10
};

export class FileTracker {
  private statecli: StateCLI;
  private config: FileTrackerConfig;
  private fileHashes: Map<string, string> = new Map();
  private changeCount: number = 0;
  private watchers: fs.FSWatcher[] = [];

  constructor(statecli: StateCLI, config?: Partial<FileTrackerConfig>) {
    this.statecli = statecli;
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  /**
   * Start watching files for changes
   */
  startWatching(): void {
    for (const watchPath of this.config.watchPaths) {
      this.watchDirectory(watchPath);
    }
  }

  /**
   * Stop watching files
   */
  stopWatching(): void {
    for (const watcher of this.watchers) {
      watcher.close();
    }
    this.watchers = [];
  }

  /**
   * Track a file change manually
   */
  trackFileChange(
    filePath: string,
    operation: 'create' | 'modify' | 'delete',
    beforeContent: string | null,
    afterContent: string | null,
    actor: string = 'ai-agent'
  ): FileChange {
    const diff = this.computeDiff(beforeContent, afterContent);
    const timestamp = new Date().toISOString();

    const change: FileChange = {
      filePath: path.normalize(filePath),
      operation,
      before: beforeContent,
      after: afterContent,
      diff,
      timestamp
    };

    // Track in StateCLI
    this.statecli.track('file', filePath, {
      operation,
      contentHash: afterContent ? this.hashContent(afterContent) : null,
      lineCount: afterContent ? afterContent.split('\n').length : 0,
      diff: diff.slice(0, 50), // Store first 50 diff lines
      timestamp
    }, actor);

    this.changeCount++;

    // Auto-checkpoint if threshold reached
    if (this.config.autoCheckpoint && 
        this.changeCount >= this.config.checkpointThreshold) {
      this.createAutoCheckpoint();
    }

    return change;
  }

  /**
   * Track a file edit with before/after content
   */
  trackEdit(
    filePath: string,
    beforeContent: string,
    afterContent: string,
    actor: string = 'ai-agent'
  ): FileChange {
    return this.trackFileChange(filePath, 'modify', beforeContent, afterContent, actor);
  }

  /**
   * Track a file creation
   */
  trackCreate(
    filePath: string,
    content: string,
    actor: string = 'ai-agent'
  ): FileChange {
    return this.trackFileChange(filePath, 'create', null, content, actor);
  }

  /**
   * Track a file deletion
   */
  trackDelete(
    filePath: string,
    previousContent: string,
    actor: string = 'ai-agent'
  ): FileChange {
    return this.trackFileChange(filePath, 'delete', previousContent, null, actor);
  }

  /**
   * Get file change history
   */
  getFileHistory(filePath: string): ReturnType<StateCLI['replay']> {
    return this.statecli.replay(`file:${path.normalize(filePath)}`);
  }

  /**
   * Compute simple diff between two strings
   */
  private computeDiff(before: string | null, after: string | null): string[] {
    const diff: string[] = [];
    
    if (before === null && after !== null) {
      // New file
      const lines = after.split('\n');
      lines.forEach((line, i) => {
        diff.push(`+${i + 1}: ${line}`);
      });
    } else if (before !== null && after === null) {
      // Deleted file
      const lines = before.split('\n');
      lines.forEach((line, i) => {
        diff.push(`-${i + 1}: ${line}`);
      });
    } else if (before !== null && after !== null) {
      // Modified file - simple line-by-line diff
      const beforeLines = before.split('\n');
      const afterLines = after.split('\n');
      
      const maxLines = Math.max(beforeLines.length, afterLines.length);
      
      for (let i = 0; i < maxLines; i++) {
        const beforeLine = beforeLines[i];
        const afterLine = afterLines[i];
        
        if (beforeLine !== afterLine) {
          if (beforeLine !== undefined) {
            diff.push(`-${i + 1}: ${beforeLine}`);
          }
          if (afterLine !== undefined) {
            diff.push(`+${i + 1}: ${afterLine}`);
          }
        }
      }
    }
    
    return diff;
  }

  /**
   * Create a hash of file content
   */
  private hashContent(content: string): string {
    return crypto.createHash('md5').update(content).digest('hex');
  }

  /**
   * Watch a directory for changes
   */
  private watchDirectory(dirPath: string): void {
    try {
      const watcher = fs.watch(dirPath, { recursive: true }, (eventType, filename) => {
        if (filename && !this.shouldIgnore(filename)) {
          this.handleFileEvent(path.join(dirPath, filename), eventType);
        }
      });
      this.watchers.push(watcher);
    } catch (error) {
      console.error(`Failed to watch directory ${dirPath}:`, error);
    }
  }

  /**
   * Check if a file should be ignored
   */
  private shouldIgnore(filePath: string): boolean {
    for (const pattern of this.config.ignorePatterns) {
      if (pattern.includes('*')) {
        const regex = new RegExp(pattern.replace('*', '.*'));
        if (regex.test(filePath)) return true;
      } else {
        if (filePath.includes(pattern)) return true;
      }
    }
    return false;
  }

  /**
   * Handle a file system event
   */
  private handleFileEvent(filePath: string, eventType: string): void {
    const normalizedPath = path.normalize(filePath);
    const previousHash = this.fileHashes.get(normalizedPath);

    try {
      if (fs.existsSync(filePath)) {
        const content = fs.readFileSync(filePath, 'utf-8');
        const newHash = this.hashContent(content);

        if (previousHash !== newHash) {
          this.fileHashes.set(normalizedPath, newHash);
          
          if (previousHash === undefined) {
            // New file or first time seeing it
            this.trackCreate(normalizedPath, content);
          } else {
            // Modified file
            this.trackEdit(normalizedPath, '', content); // We don't have old content in watch mode
          }
        }
      } else if (previousHash !== undefined) {
        // File was deleted
        this.fileHashes.delete(normalizedPath);
        this.trackDelete(normalizedPath, '');
      }
    } catch (error) {
      // File might be locked or inaccessible
    }
  }

  /**
   * Create an automatic checkpoint
   */
  private createAutoCheckpoint(): void {
    const checkpointName = `auto-${new Date().toISOString().replace(/[:.]/g, '-')}`;
    this.statecli.checkpoint('session:current', checkpointName);
    this.changeCount = 0;
  }
}
